docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -0,0 +1,597 @@
+
+_Solution_: If you need to register a synthetic bean then use the `SyntheticBeanBuildItem`.
+
+.`SyntheticBeanBuildItem` Example 1
+[source,java]
+----
+@BuildStep
+SyntheticBeanBuildItem syntheticBean() {
+   return SyntheticBeanBuildItem.configure(String.class)
+             .qualifiers(new MyQualifierLiteral())
+             .creator(mc -> mc.returnValue(mc.load("foo"))) <1>
+             .done();
+}
+----
+<1> Generate the bytecode of the `javax.enterprise.context.spi.Contextual#create(CreationalContext<T>)` implementation.
+
+The output of a bean configurator is recorded as bytecode. 
+Therefore, there are some limitations in how a synthetic bean instance is created at runtime. 
+You can:
+
+1. Generate the bytecode directly via `ExtendedBeanConfigurator.creator(Consumer<MethodCreator>)`.
+2. Pass a `io.quarkus.arc.BeanCreator` implementation via `ExtendedBeanConfiguratorio#creator(Class<? extends BeanCreator<U>>)`, and possibly specify some parameters via `ExtendedBeanConfigurator#param()`.
+3. Produce the runtime instance through a proxy returned from a <<writing-extensions.adoc#bytecode-recording,`@Recorder`>> method and set it via `ExtendedBeanConfigurator#runtimeValue(RuntimeValue<?>)` or `ExtendedBeanConfigurator#supplier(Supplier<?>)`
+
+.`SyntheticBeanBuildItem` Example 2
+[source,java]
+----
+@BuildStep
+@Record(STATIC_INIT) <1>
+SyntheticBeanBuildItem syntheticBean(TestRecorder recorder) {
+   return SyntheticBeanBuildItem.configure(Foo.class).scope(Singleton.class)
+                .runtimeValue(recorder.createFoo()) <2>
+                .done();
+}
+----
+<1> By default, a synthetic bean is initialized during `STATIC_INIT`.
+<2> The bean instance is supplied by a value returned from a recorder method.
+
+It is possible to mark a synthetic bean to be initialized during `RUNTIME_INIT`:
+
+.`RUNTIME_INIT` `SyntheticBeanBuildItem` Example
+[source,java]
+----
+@BuildStep
+@Record(RUNTIME_INIT) <1>
+SyntheticBeanBuildItem syntheticBean(TestRecorder recorder) {
+   return SyntheticBeanBuildItem.configure(Foo.class).scope(Singleton.class)
+                .setRuntimeInit() <2>
+                .runtimeValue(recorder.createFoo())
+                .done();
+}
+----
+<1> The recorder must be executed in the `ExecutionTime.RUNTIME_INIT` phase.
+<2> The bean instance is initialized during `RUNTIME_INIT`.
+
+[IMPORTANT]
+====
+Synthetic beans initialized during `RUNTIME_INIT` must not be accessed during `STATIC_INIT`. `RUNTIME_INIT` build steps that access a runtime-init synthetic bean should consume the `SyntheticBeansRuntimeInitBuildItem`:
+
+[source,java]
+----
+@BuildStep
+@Record(RUNTIME_INIT) 
+@Consume(SyntheticBeansRuntimeInitBuildItem.class) <1>
+void accessFoo(TestRecorder recorder) {
+   recorder.foo(); <2>
+}
+----
+<1> This build step must be executed after `syntheticBean()` completes.
+<2> This recorder method results in an invocation of the `Foo` bean instance.
```

Comment:
"invocation of the `Foo` bean instance" -- not sure what that means, perhaps you meant "creation" of the bean instance?

Reply:
Well, "creation" precedes "invocation" ;-). It means that inside the recorder method we do invoke a method upon the `Foo` instance of the bean registered in the snippet above.

Reply:
Ah so it should be "results in an invocation _on"_ the `Foo` bean instance" I think.